### PR TITLE
Add `fftshift` & `ifftshift` | Also migrate to GitHub CI

### LINKED ISSUE #43: Support for FFT Shifts | `fftshifted` & `ifftshifted`
Hey,

is there any possibility that we can include something like a `ShiftedArray.fftshift` or `ShiftedArray.ifftshift`.
It would be a convenient wrapper to call `fftshift` and `ifftshift` but prevent the allocation of memory.

There might be some performance issues when accidentally providing a `ShiftedArrray` to a `fft` but I'm not sure and we might discuss such issues?

If you support that idea we can help adding that :)

Thanks,

Felix

### PR DESCRIPTION
**Pull request by roflmaostc**
Hey,

as discussed, this would fix #43 . I also added tests, it seems to be fine.

Furthermore I switched to GitHub Actions, due to the changed pricing for Travis I believe that GitHub Actions is now more or less standard.

Would be happy about your feedback!

Thanks,

Felix

### CODE REVIEW

File: .codecov.yml [hunk 1]
```diff
@@ -1 +0,0 @@
-comment: false
```
Comment (piever) [inline .codecov.yml:1]: I think there is no reason to delete this, I imagine even on github CI we should still do codecov.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    tags: '*'
+jobs:
+  test:
+    name: Julia ${{ matrix.version }} - ${{ matrix.os }} - ${{ matrix.arch }} - ${{ github.event_name }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        version:
+          - '1.3' # Replace this with the minimum Julia version that your package supports. E.g. if your package requires Julia 1.5 or higher, change this to '1.5'.
```
Comment (piever) [inline .github/workflows/ci.yml:18]:
```suggestion
          - '1.0' # Replace this with the minimum Julia version that your package supports. E.g. if your package requires Julia 1.5 or higher, change this to '1.5'.
```

File: src/fft_shifts.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+"""
+function ft_center_diff(s::NTuple{N, T}, dims=ntuple(identity, Val(N))) where {N, T}
+    return ntuple(i -> i ∈ dims ?  s[i] ÷ 2 : 0 , N)
+end
+
+
+"""
+    fftshift(x [, dims])
+
+Result is semantically equivalent to `fftshift(A, dims)` but returns
```
Comment (piever) [inline src/fft_shifts.jl:26]:
```suggestion
Result is semantically equivalent to `AbstractFFTs.fftshift(A, dims)` but returns 
```

File: src/fft_shifts.jl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+```
+"""
+function fftshift(x::AbstractArray{T, N}, dims=ntuple(identity, Val(N))) where {T, N}
+    ShiftedArrays.circshift(x, ft_center_diff(size(x), dims))
+end
+
+"""
+    ifftshift(A [, dims])
+
+Result is semantically equivalent to `ifftshift(A, dims)` but returns
```
Comment (piever) [inline src/fft_shifts.jl:55]:
```suggestion
Result is semantically equivalent to `AbstractFFTs.ifftshift(A, dims)` but returns 
```

File: Project.toml [hunk 5]
```diff
@@ -3,13 +3,13 @@ uuid = "1277b4bf-5013-50f5-be3d-901d8477a67a"
...
 version = "1.0.0"
 
-[deps]
 
 [compat]
 julia = "1"
 
 [extras]
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+FFTW = "7a1cc6ca-52ef-59f5-83cd-3a7055c09341"
```
Comment (piever) [inline Project.toml:12]: Could we only add AbstractFFTs on testing? FFTW comes with some binary dependency that I'd prefer to avoid.

**Review by piever — CHANGES_REQUESTED** [review]

**Comment by piever** [comment]
Thank you, it mostly looks good, I've added a few comments.

File: .codecov.yml (hunk 1 above)
Comment (roflmaostc) [reply .codecov.yml:1]: Sure, codecov is fine, probably too many files deleted :upside_down_face:

**Comment by roflmaostc** [comment]
I added your suggestions and also added the docstrings to the docs

File: docs/src/api.md [hunk 6]
```diff
@@ -17,6 +17,14 @@ lead
 ShiftedArrays.circshift
 ```
 
+## FFT shifts
```
Comment (piever) [inline docs/src/api.md:20]:
```suggestion
## FFT shifts

```

File: docs/src/api.md [hunk 7]
```diff
@@ -17,6 +17,14 @@ lead
...
+```@docs
+ShiftedArrays.fftshift
+ShiftedArrays.ifftshift
+```
+
+
+
```
Comment (piever) [inline docs/src/api.md:27]:
```suggestion

```

File: src/fft_shifts.jl [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2,3)) # Fourier center is at (2,3,4)
+(2, 2, 3)
+```
+
+"""
+function ft_center_diff(s::NTuple{N, T}, dims=ntuple(identity, Val(N))) where {N, T}
+    return ntuple(i -> i ∈ dims ?  s[i] ÷ 2 : 0 , N)
+end
+
+
```
Comment (piever) [inline src/fft_shifts.jl:22]:
```suggestion

```

File: src/fft_shifts.jl [hunk 9]
```diff
@@ -0,0 +1,80 @@
+"""
+    ft_center_diff(s [, dims])
+
+Calculates how much the entries in each dimension must be shifted so that the
+center frequency is at the Fourier center.
+This function is internally used by `ShiftedArrays.fftshift` and `ShiftedArrays.ifftshift`
```
Comment (piever) [inline src/fft_shifts.jl:6]:
```suggestion
This function is internally used by `ShiftedArrays.fftshift` and `ShiftedArrays.ifftshift`.
```

File: test/runtests.jl [hunk 10]
```diff
@@ -79,6 +80,33 @@ end
...
+    test_fftshift(randn((10, 11)), (1,))
+    test_fftshift(randn((10, 11)), (2,))
+    test_fftshift(randn(ComplexF32,(10, 11)), (1,2))
+    test_fftshift(randn((10, 11)))
+    
+    test_fftshift(randn((10, 11, 12, 13)), (2,4))
+    test_fftshift(randn((10, 11, 12, 13)), (5))
+    test_fftshift(randn((10, 11, 12, 13)))
+   
+
```
Comment (piever) [inline test/runtests.jl:103]:
```suggestion
   
```

File: test/runtests.jl [hunk 11]
```diff
@@ -79,6 +80,33 @@ end
...
+    @test (2,2, 0) == ShiftedArrays.ft_center_diff((4,5, 6), (1,2)) # Fourier center is at (2,3,0)
+    @test (2,2,3) == ShiftedArrays.ft_center_diff((4,5, 6), (1,2,3)) # Fourier center is at (2,3,4)
```
Comment (piever) [inline test/runtests.jl:105]:
```suggestion
    @test (2, 2, 0) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2)) # Fourier center is at (2, 3, 0)
    @test (2, 2, 3) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2, 3)) # Fourier center is at (2, 3, 4)
```

File: test/runtests.jl [hunk 12]
```diff
@@ -79,6 +80,33 @@ end
...
+end
+
+
+
```
Comment (piever) [inline test/runtests.jl:109]:
```suggestion

```

File: docs/src/api.md [hunk 13]
```diff
@@ -17,6 +17,12 @@ lead
 ShiftedArrays.circshift
 ```
 
+## FFT shifts
+
+```@docs
+ShiftedArrays.fftshift
+ShiftedArrays.ifftshift
+```
```
Comment (piever) [inline docs/src/api.md:25]:
```suggestion
```

```

File: docs/src/api.md [hunk 14]
```diff
@@ -17,6 +17,12 @@ lead
...
 ## Accessor functions
```
Comment (piever) [inline docs/src/api.md:26]:
```suggestion

## Accessor functions
```

File: test/runtests.jl [hunk 15]
```diff
@@ -79,6 +80,30 @@ end
...
+    test_fftshift(randn((10, 11)), (2,))
+    test_fftshift(randn(ComplexF32,(10, 11)), (1,2))
+    test_fftshift(randn((10, 11)))
+    
+    test_fftshift(randn((10, 11, 12, 13)), (2,4))
+    test_fftshift(randn((10, 11, 12, 13)), (5))
+    test_fftshift(randn((10, 11, 12, 13)))
+   
+
+    @test (2, 2, 0) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2)) # Fourier center is at (2, 3, 0)
```
Comment (piever) [inline test/runtests.jl:104]:
```suggestion
    end

    test_fftshift(randn((10,)))
    test_fftshift(randn((11,)))
    test_fftshift(randn((10,)), (1,))
    test_fftshift(randn(ComplexF32, (11,)), (1,))
    test_fftshift(randn((10, 11)), (1,))
    test_fftshift(randn((10, 11)), (2,))
    test_fftshift(randn(ComplexF32,(10, 11)), (1,2))
    test_fftshift(randn((10, 11)))

    test_fftshift(randn((10, 11, 12, 13)), (2,4))
    test_fftshift(randn((10, 11, 12, 13)), (5))
    test_fftshift(randn((10, 11, 12, 13)))

    @test (2, 2, 0) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2)) # Fourier center is at (2, 3, 0)
```

File: test/runtests.jl [hunk 16]
```diff
@@ -79,6 +80,29 @@ end
     @test all(circshift(v, 3) .== ShiftedArrays.circshift(v, 3))
 end
 
+
+@testset "fftshift and ifftshift" begin
```
Comment (piever) [inline test/runtests.jl:84]:
```suggestion
end

@testset "fftshift and ifftshift" begin
```

File: test/runtests.jl [hunk 17]
```diff
@@ -79,6 +80,29 @@ end
...
+    test_fftshift(randn((10, 11)))
+
+    test_fftshift(randn((10, 11, 12, 13)), (2,4))
+    test_fftshift(randn((10, 11, 12, 13)), (5))
+    test_fftshift(randn((10, 11, 12, 13)))
+
+    @test (2, 2, 0) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2)) # Fourier center is at (2, 3, 0)
+    @test (2, 2, 3) == ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2, 3)) # Fourier center is at (2, 3, 4)
+end
 @testset "laglead" begin
```
Comment (piever) [inline test/runtests.jl:106]:
```suggestion
end

@testset "laglead" begin
```

File: src/fft_shifts.jl [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+# Examples
+```jldoctest
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2)) # Fourier center is at (2,3,0)
+(2, 2, 0)
+
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2,3)) # Fourier center is at (2,3,4)
+(2, 2, 3)
+```
+
```
Comment (piever) [inline src/fft_shifts.jl:16]:
```suggestion
```

File: src/fft_shifts.jl [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+ 0  0  0
+ 0  0  0
+
+julia> ShiftedArrays.ifftshift([0 1 0; 0 0 0; 0 0 0], (2,))
+3×3 CircShiftedArray{Int64, 2, Matrix{Int64}}:
+ 1  0  0
+ 0  0  0
+ 0  0  0
+```
+
```
Comment (piever) [inline src/fft_shifts.jl:74]:
```suggestion
```

File: src/fft_shifts.jl [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+(2, 2, 0)
+
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2,3)) # Fourier center is at (2,3,4)
+(2, 2, 3)
+```
+
+"""
+function ft_center_diff(s::NTuple{N, T}, dims=ntuple(identity, Val(N))) where {N, T}
+    return ntuple(i -> i ∈ dims ?  s[i] ÷ 2 : 0 , N)
+end
```
Comment (piever) [inline src/fft_shifts.jl:20]:
```suggestion
end

```

File: src/fft_shifts.jl [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+center frequency is at the Fourier center.
+This function is internally used by `ShiftedArrays.fftshift` and `ShiftedArrays.ifftshift`.
+
+# Examples
+```jldoctest
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2)) # Fourier center is at (2,3,0)
+(2, 2, 0)
+
+julia> ShiftedArrays.ft_center_diff((4,5, 6), (1,2,3)) # Fourier center is at (2,3,4)
+(2, 2, 3)
```
Comment (piever) [inline src/fft_shifts.jl:14]:
```suggestion
julia> ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2)) # Fourier center is at (2, 3, 0)
(2, 2, 0)

julia> ShiftedArrays.ft_center_diff((4, 5, 6), (1, 2, 3)) # Fourier center is at (2, 3, 4)
(2, 2, 3)
```

File: src/fft_shifts.jl [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+julia> ShiftedArrays.ifftshift([0 1 0; 0 0 0; 0 0 0], (2,))
+3×3 CircShiftedArray{Int64, 2, Matrix{Int64}}:
+ 1  0  0
+ 0  0  0
+ 0  0  0
+```
+"""
+function ifftshift(x::AbstractArray{T, N}, dims=ntuple(identity, Val(N))) where {T, N}
+    ShiftedArrays.circshift(x, .-(ft_center_diff(size(x), dims)))
```
Comment (piever) [inline src/fft_shifts.jl:76]:
```suggestion
    ShiftedArrays.circshift(x, map(-, ft_center_diff(size(x), dims)))
```

File: src/fft_shifts.jl [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    return ntuple(i -> i ∈ dims ?  s[i] ÷ 2 : 0 , N)
+end
+
+"""
+    fftshift(x [, dims])
+
+Result is semantically equivalent to `AbstractFFTs.fftshift(A, dims)` but returns 
+a `CircShiftedArray` instead. 
+
+# Examples
```
Comment (piever) [inline src/fft_shifts.jl:27]:
```suggestion
# Examples

```

File: src/fft_shifts.jl [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    ShiftedArrays.circshift(x, ft_center_diff(size(x), dims))
+end
+
+"""
+    ifftshift(A [, dims])
+
+Result is semantically equivalent to `AbstractFFTs.ifftshift(A, dims)` but returns 
+a `CircShiftedArray` instead. 
+
+# Examples
```
Comment (piever) [inline src/fft_shifts.jl:56]:
```suggestion
# Examples

```

File: src/fft_shifts.jl [hunk 25]
```diff
@@ -0,0 +1,77 @@
+"""
+    ft_center_diff(s [, dims])
+
+Calculates how much the entries in each dimension must be shifted so that the
+center frequency is at the Fourier center.
+This function is internally used by `ShiftedArrays.fftshift` and `ShiftedArrays.ifftshift`.
+
+# Examples
```
Comment (piever) [inline src/fft_shifts.jl:8]:
```suggestion
# Examples

```

**Review by piever — APPROVED** [review]

**Comment by piever** [comment]
Thanks. I've fixed the formatting (it should be uniform with the rest of the package) and merged.

**Pull request merged**